### PR TITLE
Prohibit duplicate function declarations

### PR DESCRIPTION
**Pull request by adamretter**
Prohibit functions with the same name and arity as per the XQuery spec.

Closes https://github.com/eXist-db/exist/issues/1665
Closes https://github.com/eXist-db/exist/issues/1174

### CODE REVIEW

File: exist-core/src/test/xquery/xquery3/arrowop.xql [hunk 1]
```diff
@@ -237,14 +237,14 @@ function ao:fold-left-with-contextitem () {
...
-function ao:fold-right-with-contextitem () {
+function ao:fold-right-with-contextitem-1 () {
     (<a/>,<b/>,<c/>,<d/>)
         => fold-right('', function ($a, $r) { $r || xs:string(node-name($a)) })
 };
 
 declare
     %test:assertEquals('dcba')
-function ao:fold-right-with-contextitem () {
+function ao:fold-right-with-contextitem-2 () {
```
Comment (line-o) [inline exist-core/src/test/xquery/xquery3/arrowop.xql:247]: I see what you found here :)

**Comment by line-o** [comment]
This is a very nice bugfix. I am cautious to approve it only because I cannot know implications of:

- Object2ObjectRBTreeMap
- Object2ObjectAVLMap
- when combining hashes the first part is multiplied by 31

I hope that @adamretter you could either explain why you chose those or somebody who understands approves.

**Comment by adamretter** [comment]
@line-o to explain:

> I am cautious to approve it only because I cannot know implications of:
>
> * Object2ObjectRBTreeMap
> * Object2ObjectAVLMap

These are alternatives to Java's TreeMap. They should offer better performance and lower memory-use. I chose them based on their performance characteristics and our needs. You can read more about the performance trade-offs here - http://fastutil.di.unimi.it/docs/#Performance

> *when combining hashes the first part is multiplied by 31

This is kinda standard stuff for hashes... 31 is a prime number ;-)

**Comment by line-o** [comment]
Thanks for the explanation. I found this magic prime in a long discussion over at stackoverflow.
I would love to hear why the o2o by-reference binary tree is used in one place and the AVL one in another, but I will use this as a CS exercise for later.

**Comment by line-o** [comment]
When I have time and the tests pass locally I might merge before Travis is fixed... whatever comes first.

**Comment by line-o** [comment]
So just for reference:

- RBTree (red-black tree) is good for Maps with frequent adds and deletions
- AVLTree is better for frequent searches and less frequent additions and deletions

So `declaredFunctions` seems to be a perfect fit for an AVLTree ... things starting to make sense

**Comment by line-o** [comment]
@adamretter are additions and deletions so frequent for global variables?

**Comment by line-o** [comment]
@adamretter I would like to know why you chose RBTree over AVLTree for global variables.

**Comment by adamretter** [comment]
@line-o From the fastutil doc:

> So on small trees red-black trees could be faster, but on very large sets AVL trees will shine

I figured that there would be very few global variables generally.

**Comment by line-o** [comment]
Thanks for the clarification. Without data we can only guess. 
It is very likely, that there will be less global variables than functions declared. It would be helpful to know where the boundary is where AVLTrees start to be faster. If we are talking millions, then RBTree is better for functions, too.

**Comment by line-o** [comment]
Both trees will never exceed 10K (another informed guess)

**Comment by adamretter** [comment]
I doubt they would exceed 1K

On Sun, 12 Apr 2020 at 16:14, Juri Leino <notifications@github.com> wrote:
>
> Both trees will never exceed 10K (another informed guess)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.



-- 
Adam Retter

skype: adam.retter
tweet: adamretter
http://www.adamretter.org.uk

**Comment by line-o** [comment]
Is 1K very large? Large enough to benefit from AVLTree?

**Comment by adamretter** [comment]
@line-o No. But it is larger than what I thought would be small for the RBTree. Maybe better to discuss by chat if you still have questions

**Comment by line-o** [comment]
The implementation of Java HashMap uses RBTrees since ... wait for it ... version 8 
source: http://coding-geek.com/how-does-a-hashmap-work-in-java/#JAVA_8_improvements

**Comment by adamretter** [comment]
@line-o That may be the case, but they are still more memory hungry than fastutil

**Pull request merged**